### PR TITLE
Make privacy visitor use types more (instead of HIR)

### PR DESCRIPTION
**Pull request by oli-obk**
r? @petrochenkov 

This is a prerequisite to normalizing projections, as otherwise we have too many invalid bound vars (hir_ty_to_ty is creating types that have bound vars, but no binder).

The commits are still chaotic, I'm gonna clean them up, but I just wanted to let you know about the general direction and wondering if we could land this before adding normalization, as normalization is where behavioral changes happen, and I'd like to keep that part as minimal as possible.

[context can be found on zulip](https://rust-lang.zulipchat.com/#narrow/stream/315482-t-compiler.2Fetc.2Fopaque-types/topic/weak.20type.20aliases.20and.20privacy)

### CODE REVIEW

**Comment by petrochenkov** [comment]
cc @Bryanskiy

File: compiler/rustc_privacy/src/lib.rs [hunk 1]
```diff
@@ -1347,16 +1389,34 @@ impl<'tcx> Visitor<'tcx> for TypePrivacyVisitor<'tcx> {
     fn visit_item(&mut self, item: &'tcx hir::Item<'tcx>) {
         let orig_current_item = mem::replace(&mut self.current_item, item.owner_id.def_id);
         let old_maybe_typeck_results = self.maybe_typeck_results.take();
-        intravisit::walk_item(self, item);
+        rustc_ty_utils::sig_types::walk_types(self.tcx(), item.owner_id.def_id, self);
```
[the remark this answers is not in the document]
Comment (petrochenkov) [reply compiler/rustc_privacy/src/lib.rs:1392]: I also don't understand how `walk_types` checks types in signatures, and why the additional code below is necessary.

If we have a, for example, function
```
// Assume `Alias` always results in `u8`, regardless of type arguments
fn foo(arg: Alias<Priv, Pub>) {}
```
we need to recursively walk HIR types in its signature, normalize them, and check for privacy, i.e.

- `Alias<Priv, Pub>` -> normalize -> `u8` -> public, ok
- `Priv` -> normalize -> `Priv` -> private, error
- `Pub` -> normalize -> `Pub` -> public, ok

(Setting good spans important, but it wasn't the primary point of visiting HIR here.)

`walk_types` seems to only take the top level type of `arg`, and then use `TypeVisitable` on it (i.e. on `u8`) and ignore `Priv` and `Pub`.
I added tests for cases like this back in 2017, not sure why they don't fail.

**Comment by petrochenkov** [comment]
(I also don't understand all the new binder play here, but I assume it's not something complex, and you know what you are doing.)

**Comment by petrochenkov** [comment]
>wondering if we could land this before adding normalization, as normalization is where behavioral changes happen, and I'd like to keep that part as minimal as possible.

Yes, that's quite reasonable.

File: compiler/rustc_privacy/src/lib.rs [hunk 2]
```diff
@@ -176,6 +192,20 @@ where
...
+        t: &ty::Binder<'tcx, T>,
+    ) -> ControlFlow<Self::BreakTy> {
+        self.universes.push(None);
+        let result: ControlFlow<<V as DefIdVisitor<'tcx>>::BreakTy> = t.super_visit_with(self);
+        self.universes.pop();
+        result
+    }
+
+    fn visit_predicate(&mut self, p: ty::Predicate<'tcx>) -> ControlFlow<Self::BreakTy> {
+        self.visit_clause(p.as_clause().unwrap())
```
[the remark this answers is not in the document]
Comment (oli-obk) [reply compiler/rustc_privacy/src/lib.rs:206]: The visit_clause has special logic for traits. This type visitor would not use it and thus only look at the types in the trait ref arguments, instead of discovering the trait ref's trait and marking it('s def id) as seen

File: compiler/rustc_privacy/src/lib.rs [hunk 3]
```diff
@@ -356,10 +380,11 @@ trait VisibilityLike: Sized {
         tcx: TyCtxt<'_>,
         effective_visibilities: &EffectiveVisibilities,
     ) -> Self {
-        let mut find = FindMin::<_, SHALLOW> { tcx, effective_visibilities, min: Self::MAX };
+        let infcx = tcx.infer_ctxt().build();
```
[the remark this answers is not in the document]
Comment (oli-obk) [reply compiler/rustc_privacy/src/lib.rs:383]: We could use a single inferctxt always. Or generate it whenever we create the universe table. I'll convert to doing the latter. This inferctxt logic is leftover from a previous version of this PR

File: compiler/rustc_privacy/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (oli-obk) [reply compiler/rustc_privacy/src/lib.rs:1392]: Walk_types does not normalize. When we start normalizing, we'll run on both the normalized type and the unnormalized type to catch everything as you described

**Comment by oli-obk** [comment]
> (I also don't understand all the new binder play here, but I assume it's not something complex, and you know what you are doing.)

I'll pull it into a separate commit. It's not necessary until we do normalization, and it's indeed just a mechanical refactoring to not lose the binders

File: compiler/rustc_privacy/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (oli-obk) [reply compiler/rustc_privacy/src/lib.rs:1392]: > Now they are only check in signatures of free items, as I understand.

Yea I was wondering about that. I'll add some tests

File: compiler/rustc_privacy/src/lib.rs [hunk 4]
```diff
@@ -232,11 +247,15 @@ where
...
-                    data.def_id,
-                    "associated type",
-                    &LazyDefPathStr { def_id: data.def_id, tcx },
-                )?;
+                let (def_id, kind) = match kind {
+                    ty::Inherent => (data.def_id, "associated type"),
+                    ty::Weak => (data.def_id, "type alias"),
+                    // Trait associated types don't have visibility, they are
+                    // visible if their parent trait is.
+                    ty::Projection => (tcx.parent(data.def_id), "trait"),
```
[the remark this answers is not in the document]
Comment (oli-obk) [reply compiler/rustc_privacy/src/lib.rs]: d9726ca76df26718ce40cfb4d2eaf9b7076e0226 shows why we need it

File: compiler/rustc_privacy/src/lib.rs [hunk 5]
```diff
@@ -207,22 +237,7 @@ where
...
-                if V::SKIP_ASSOC_TYS {
-                    // Visitors searching for minimal visibility/reachability want to
-                    // conservatively approximate associated types like `<Type as Trait>::Alias`
-                    // as visible/reachable even if both `Type` and `Trait` are private.
-                    // Ideally, associated types should be substituted in the same way as
-                    // free type aliases, but this isn't done yet.
-                    return ControlFlow::Continue(());
-                }
-                // This will also visit substs if necessary, so we don't need to recurse.
-                return self.visit_projection_ty(proj);
```
[the remark this answers is not in the document]
Comment (oli-obk) [reply compiler/rustc_privacy/src/lib.rs:223]: yes

File: compiler/rustc_privacy/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (oli-obk) [reply compiler/rustc_privacy/src/lib.rs]: I just removed it for now, we only need it when we actually normalize

File: compiler/rustc_privacy/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (oli-obk) [reply compiler/rustc_privacy/src/lib.rs:1359]: We have all the tests we could want. There's also `PrivateItemsInPublicInterfacesChecker` which does something similar, but apparently not exactly the same thing? I'm confused 😆

[the remark this answers is not in the document]
Comment (oli-obk) [reply compiler/rustc_privacy/src/lib.rs:1359]: It's not clear to me what `check_mod_privacy` and `check_private_in_public` are supposed to do differently and why they do a lot of the same things.

[the remark this answers is not in the document]
Comment (petrochenkov) [reply compiler/rustc_privacy/src/lib.rs:1359]: Briefly, `check_mod_privacy` checks uses, it's precise (report errors) and pretty conservative / future-proofing (e.g. it's not entirely clear whether we *should* report errors for `Priv` in https://github.com/rust-lang/rust/pull/113671#discussion_r1263847616).

`check_private_in_public` checks definitions trying to predict which definitions will later result in `check_mod_privacy` errors on use, it's imprecise (reports lints) and misses some cases (allow leaking private stuff).

There's also some legacy stuff in there.
Everything will become a bit simpler when https://github.com/rust-lang/rust/pull/113126 lands, if you want we can delay this PR until then.

Further simplifications / deduplication may also be unlocked by correctly normalizing all associated types.

**Comment by petrochenkov** [comment]
https://github.com/rust-lang/rust/pull/113126 has landed.
@rustbot author

**Comment by oli-obk** [comment]
@rustbot ready

not sure about the last commit, but I added it to show why I'm special casing projections

File: compiler/rustc_privacy/src/lib.rs [hunk 6]
```diff
@@ -232,11 +221,21 @@ where
...
-                    data.def_id,
-                    "associated type",
-                    &LazyDefPathStr { def_id: data.def_id, tcx },
-                )?;
+                let kind = match kind {
+                    ty::Inherent => Some("associated type"),
+                    ty::Weak => Some("type alias"),
+                    // Trait associated types don't have visibility, they are
+                    // visible if their parent trait is.
+                    ty::Projection => None,
```
Comment (petrochenkov) [inline compiler/rustc_privacy/src/lib.rs:229]:
```suggestion
                    ty::Projection => Some("associated type"),
```
Not sure what the last commit intends to demonstrate.
I didn't suggest skipping this logic for associated types, I only suggested to use their own def ids.

Associated type def ids do have (semantic) visibility (the comment above is wrong in this regard), but that visibility is indeed inherited from the trait (the comment above is right in that regard).

Comment (oli-obk) [reply compiler/rustc_privacy/src/lib.rs]: yea that makes sense. I added a change that just exposes associated types directly here. I will edit git history at the end before merging the PR

**Comment by oli-obk** [comment]
This is ready for review again

**Comment by wesleywiser** [comment]
> I still have other questions, but it would be good to sort out the last commit first.

@petrochenkov it's not clear to me, did those questions get resolved or are there still open threads of discussion? 

Thanks!

**Comment by petrochenkov** [comment]
I was just busy for the most of September and accumulated a large review queue.
This PR is the last in the queue and I'm actually reviewing it today.

File: tests/ui/privacy/private-in-public.rs [hunk 7]
```diff
@@ -104,8 +104,8 @@ mod aliases_pub {
 
     // This should be OK, but associated type aliases are not substituted yet
     pub fn f3(arg: <Priv as PrivTr>::Assoc) {}
-    //~^ WARNING trait `aliases_pub::PrivTr` is more private than the item `aliases_pub::f3`
-    //~| WARNING type `aliases_pub::Priv` is more private than the item `aliases_pub::f3`
+    //~^ WARNING type `aliases_pub::Priv` is more private than the item `aliases_pub::f3`
```
Comment (petrochenkov) [inline tests/ui/privacy/private-in-public.rs:107]: Nit: unnecessary annotation order change here and below.

File: compiler/rustc_privacy/src/lib.rs [hunk 8]
```diff
@@ -1193,7 +1196,7 @@ impl<'tcx> Visitor<'tcx> for TypePrivacyVisitor<'tcx> {
             // Types in signatures.
             // FIXME: This is very ineffective. Ideally each HIR type should be converted
             // into a semantic type only once and the result should be cached somehow.
-            if self.visit(rustc_hir_analysis::hir_ty_to_ty(self.tcx, hir_ty)).is_break() {
+            if self.visit(rustc_hir_analysis::hir_ty_to_ty(self.tcx(), hir_ty)).is_break() {
```
Comment (petrochenkov) [inline compiler/rustc_privacy/src/lib.rs:1199]:
```suggestion
            if self.visit(rustc_hir_analysis::hir_ty_to_ty(self.tcx, hir_ty)).is_break() {
```

File: compiler/rustc_privacy/src/lib.rs [hunk 9]
```diff
@@ -1358,7 +1361,20 @@ impl<'tcx> Visitor<'tcx> for TypePrivacyVisitor<'tcx> {
...
+            | hir::ItemKind::Const(_, _, body_id)
+            | hir::ItemKind::Fn(_, _, body_id) => self.visit_nested_body(body_id),
+            hir::ItemKind::Impl(hir::Impl { of_trait: Some(tr), .. }) => {
+                self.span = tr.path.span;
+                let trait_ref =
+                    self.tcx().impl_trait_ref(item.owner_id.def_id).unwrap().instantiate_identity();
+                self.visit_def_id(trait_ref.def_id, "trait", &trait_ref.print_only_trait_path());
+            }
+            _ => {}
+        }
```
Comment (petrochenkov) [inline compiler/rustc_privacy/src/lib.rs:1377]: Wait, if `fn visit_ty` with `hir_ty_to_ty` still exists, then why this change is needed?
All the signature types will be visited by `visit_ty` anyway.

File: compiler/rustc_privacy/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (petrochenkov) [reply compiler/rustc_privacy/src/lib.rs:180]: How did this work previously?
Does this change fix some pre-existing issue?

I tried to remove this change and got some missing errors.
But without this PR, on master, those errors are reported even without the `visit_predicate` override.

**Comment by petrochenkov** [comment]
Can we merge everything besides the `fn visit_item` change in https://github.com/rust-lang/rust/pull/113671/commits/f507b14b3dbc7b4e164ed8a2a900a051ca1c79d6 first?
I'm still not sure that it's doing something right, but I'll return to it later.
@rustbot author

File: compiler/rustc_privacy/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (petrochenkov) [reply compiler/rustc_privacy/src/lib.rs:180]: FWIW, I still cannot find where the *real* default implementation of `fn visit_predicate` lives, there are too many abstraction layers.

File: compiler/rustc_ty_utils/src/sig_types.rs [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+//! a signature, while preserving spans as much as possible
+
+use std::ops::ControlFlow;
+
+use rustc_hir::{def::DefKind, def_id::LocalDefId};
+use rustc_middle::ty::{self, TyCtxt};
+use rustc_span::Span;
+use rustc_type_ir::visit::TypeVisitable;
+
+pub trait Spanned<'tcx> {
```
[the remark this answers is not in the document]
Comment (oli-obk) [reply compiler/rustc_ty_utils/src/sig_types.rs:11]: addressed in https://github.com/rust-lang/rust/pull/117076

File: compiler/rustc_ty_utils/src/opaque_types.rs [hunk 11]
```diff
@@ -273,25 +276,14 @@ fn opaque_types_defined_by<'tcx>(tcx: TyCtxt<'tcx>, item: LocalDefId) -> &'tcx [
...
-                Some(ty) => ty.span,
-                _ => tcx.def_span(item),
-            };
-            collector.visit_spanned(span, tcx.type_of(item).instantiate_identity());
+        DefKind::AssocFn
+        | DefKind::Fn
+        | DefKind::Static(_)
+        | DefKind::Const
+        | DefKind::AssocConst
+        | DefKind::AnonConst => {
```
[the remark this answers is not in the document]
Comment (oli-obk) [reply compiler/rustc_ty_utils/src/opaque_types.rs:286]: addressed in https://github.com/rust-lang/rust/pull/117076

**Comment by petrochenkov** [comment]
Blocked on https://github.com/rust-lang/rust/pull/117076.
@rustbot blocked

File: compiler/rustc_privacy/src/lib.rs (hunk 9 above)
Comment (oli-obk) [reply compiler/rustc_privacy/src/lib.rs:1337]: Oh yea that was all just dead code. I removed it

**Comment by petrochenkov** [comment]
(I'll return to this around early December, I think.)

**Comment by oli-obk** [comment]
I'm on leave until January anyway 😃

**Comment by petrochenkov** [comment]
Ok, what I see here now:
- Types in signatures are not checked in trait/impl/foreign items and in fields.
- Types are not checked before type alias normalization, i.e. type `IgnoreArguments<PrivTy>` won't be reported, but it was reported previously. (Or at least I fail to see how they could be reported.)
@rustbot author

**Comment by oli-obk** [comment]
> Types in signatures are not checked in trait/impl/foreign items and in fields.

They are now. Nothing changed, so they must have been checked before, probably by going down the HIR route in the recursive visitor. Now they are taking the ty route.

> Types are not checked before type alias normalization, i.e. type `IgnoreArguments<PrivTy>` won't be reported, but it was reported previously. (Or at least I fail to see how they could be reported.)

that's because those type aliases are prevented with hard errors: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=e143c6075b63bc855e751c7053bf2329

**Comment by petrochenkov** [comment]
>that's because those type aliases are prevented with hard errors: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=e143c6075b63bc855e751c7053bf2329

Interesting.
That's true for free type aliases, but not for associated type aliases, e.g. in this example
```rust
#![feature(type_alias_impl_trait)]
#![allow(unused)]

mod m {
    struct Priv;
    pub type Leak = Priv;
}

trait Trait {
    type A<T>;
}

impl Trait for u8 {
    type A<T> = u8;
}

fn check() -> <u8 as Trait>::A<m::Leak> {
    0
}
```

This example still produces an error on this PR's branch though, not sure why.

File: compiler/rustc_privacy/src/lib.rs [hunk 12]
```diff
@@ -1085,71 +1098,39 @@ impl<'tcx> Visitor<'tcx> for TypePrivacyVisitor<'tcx> {
...
             }
         } else {
-            span_bug!(self.span, "`hir::InferArg` outside of a body");
+            // FIXME: check types of const infers here.
         }
         intravisit::walk_inf(self, inf);
     }
 
     fn visit_trait_ref(&mut self, trait_ref: &'tcx hir::TraitRef<'tcx>) {
         self.span = trait_ref.path.span;
```
Comment (petrochenkov) [inline compiler/rustc_privacy/src/lib.rs:1132]: It's probably unnecessary to override `visit_trait_ref` now, because this assignment is dead - the span is always overwritten by a more precise span in `walk_trait_ref`.

File: compiler/rustc_privacy/src/lib.rs [hunk 13]
```diff
@@ -13,6 +13,7 @@ extern crate tracing;
 
 mod errors;
 
+use hir::ItemKind;
```
Comment (petrochenkov) [inline compiler/rustc_privacy/src/lib.rs:16]: Nit: non-standard imports was auto-inserted by an editor.

**Comment by petrochenkov** [comment]
>They are now. Nothing changed, so they must have been checked before, probably by going down the HIR route in the recursive visitor. Now they are taking the ty route.

Fields are still not visited.
This snippet doesn't error on this PR.
```rust
mod m {
    struct Priv;
    pub type Leak = Priv;
}

struct S {
    field: m::Leak
}
```

I don't like how easy it is to miss something with this approach :(
Previously `visit_ty` covered everything automatically.

**Comment by petrochenkov** [comment]
>I don't like how easy it is to miss something with this approach :(
>Previously `visit_ty` covered everything automatically.

Is it possible for `walk_types` to e.g. collect a set of visited `HirId`s, and then for `visit_ty`/`visit_trait_ref` to panic on anything that is not in that set?

**Comment by oli-obk** [comment]
> I don't like how easy it is to miss something with this approach :(
> Previously `visit_ty` covered everything automatically.

actually that is because of a change in the `walk_types` logic where we used to walk into fields, but now stopped doing that *recursively* but I also accidentally made it stop doing that on the outermost level (struct/union/enum declarations).

I don't think this change is more fragile, it's just the conversion from one scheme to another that is bug-prone.

**Comment by oli-obk** [comment]
> This example still produces an error on this PR's branch though, not sure why.

we walk the arguments of the projection. We don't normalize, so if there's a private type behind the projection then we'll miss it. Not sure that can happen, as the impl would probably be the problematic one then.

**Comment by petrochenkov** [comment]
> > This example still produces an error on this PR's branch though, not sure why.
> 
> [we walk the arguments of the projection](https://github.com/rust-lang/rust/blob/5c25de67d49014397d5c09bcac5f2770b9fda1be/compiler/rustc_privacy/src/lib.rs#L141). We don't normalize, so if there's a private type behind the projection then we'll miss it. Not sure that can happen, as the impl would probably be the problematic one then.

From what I understand, the compiler will eventually keep all type aliases in the `ty` representation.
That will eliminate the risks related to `AliasIgnoringArguments<PrivTy>` in all cases, we'll likely just need to check privacy at every "normalization step", instead of "every HIR type" like before this PR.

**Comment by petrochenkov** [comment]
Could you add the case from https://github.com/rust-lang/rust/pull/113671#issuecomment-1931903873 to tests (if there's no equivalent)?
r=me after that.
@rustbot author
